### ISSUE TITLE
Work specs for disaster recovery

### ISSUE THREAD

**Issue by zaibon** [zero-os/0-Disk#488]
See : https://github.com/zero-os/0-Disk/blob/disaster_recovery/specs/disaster_recovery.md

**Comment by GlenDC** [comment]
> To re-distribute the lost data to the remaining shards of the master cluster, the tlog will have to replay all the vdisks till all the data have been rewritten. This is needed cause the tlog doesn't have any knowledge of which block belongs to which shards.

It would be great if this would be possible in a server-filter mode. Meaning that we would only write the content during replaying for the actual servers that failed. I think right now it is not possible to only partial data from the tlogserver due to how it's stored. But that shouldn't stop us from afterwards (during the replay process, where content is already fetched, and decoded and such) to filter the content right there and then, such that we only write to previously failed servers.

If _all_ servers for a given cluster failed it would still result to the same process as described in this spec, but I think it will be much more likely that only _some_ servers will have failed completely (for both primary and slave). Rather than that happening for _all_ servers on both primary and slave at once.

However, without #483 there is no easy way to support such feature. That would mean that you:

+ either will have to ignore this comment for now;
+ or that you simply hack it in temporarily for now;
+ or, most unlikely, be blocked on this issue for now;

**Comment by iwanbk** [comment]
> (*) Well easily... Check my comment about why supporting this feature right now is not really possible yet.

@zaibon found this func https://github.com/zero-os/0-Disk/blob/6c7e5bf4c8acac254fee860cd81c08a9d2e4e400/nbd/ardb/provider.go#L136-L169

I think we can extract almost all the code of that func to find the shard number of an index.
Then if it is the failed the failed shard, we can replay the index, otherwise we can ignore it.

**Comment by GlenDC** [comment]
> I think we can extract almost all the code of that func to find the shard number of an index.
Then if it is the failed the failed shard, we can replay the index, otherwise we can ignore it.

Sure, that much is obvious. But that's not code that should be copied between modules. Instead as part of the new design it should be possible to easily get this kind of information directly from the in-memory cluster type.

**Comment by iwanbk** [comment]
>  But that's not code that should be copied between modules. Instead as part of the new design it should be possible to easily get this kind of information directly from the in-memory cluster type.

@zaibon 
I proposed this:
1) i don't copy part of the code, but make it public func, so it can be used by both  tlog player and `(rp *staticRedisProvider) DataConnection` so we have working version soon.
2) in case of @muhamadazmy or @GlenDC could refactor/redesign the ardb provider, tlog player will switch to use new code. 
3) we added new option to `zeroctl restore`, it is `failedClusterID`:
      - if exist, we write only data from failed shard
      - if not, we write all

I proposed point (1) because it seems we need to get a working version soon.
Of course the ideal way is start from point(2)
